schoenrock-rossiter_tom_ndvi.py

```diff
@@ -0,0 +1,656 @@
+# 
+# Use the functions that we reviewed in class (or create your own versions of them) to implement your code
+# 
+# ### In the Cell below Place  All Functions Needed to Run this Notebook (20 points)
+
+# In[4]:
+
+
+### DO NOT REMOVE THIS LINE OR EDIT / MOVE THIS CELL ###
+start_time = datetime.now()
+
+
+# The cell below contains the three functions that were used to streamline the amount of code and repitition used. The third function takes the previous two functions, one of which opens a sinlge band and crops it and the other takes the cropped band and masks it using a qa layer, and combines them so that they can be called in the subsequent loop. The loop applies this code to each landsat scene file for both of the sites before outputting a list.
+
+# In[5]:
+
+
+# In this cell place all of the functions needed to run your notebook
+# You will be graded here on function application, docstrings, efficiency so ensure
+# All functions are placed here!
+
+
+#Create paths to both sites
+
+
+
+site_path = os.path.join("ndvi-automation",
+                         "sites")
+
+# path to sjer and harv sites
+study_sites = glob(os.path.join(site_path + "/*/"))
+print(study_sites)
+
+#function to open single landsat band, mask using qa and crop
+def open_clean_bands(band_path,
+                     crop_extent,
+                     valid_range=None,):
+    """Open and mask a single landsat band using a pixel_qa layer.
+
+    Parameters
+    -----------
+    band_path : string
+        A path to the array to be opened
+    crop_extent : vector
+        A path to the vector file to use as crop boundary
+    valid_range : tuple (optional)
+        A tuple of min and max range of values for the data. Default = None
+
+
+    Returns
+    -----------
+    arr : xarray DataArray
+        An xarray DataArray with values that should be masked set to 1 for 
+        True (Boolean)
+    """
+    band = rxr.open_rasterio(band_path, masked = True).rio.clip    (crop_extent.geometry, from_disk = True).squeeze()
+
+    # Only run this step if a valid range tuple is provided
+    if valid_range:
+        mask = ((band < valid_range[0]) | (band > valid_range[1]))
+        band = band.where(~xr.where(mask, True, False))
+
+    return band
+
+# function to apply cloud mask and caluclate ndvi (mean)
+def mask_crop_ndvi(all_bands,
+                   crop_bound,
+                   pixel_qa,
+                   masked_values):
+    """Open and mask a single landsat band using a pixel_qa layer.
+
+    Parameters
+    -----------
+    all_bands : list
+        a list containing the xarray objects for landsat  bands 4 and  5
+    crop_bound: geopandas GeoDataFrame
+        A geopandas dataframe to be used to crop the raster data using
+        rasterio mask().
+    pixel_qa: xarray DataArray
+        An xarray DataArray with pixel qa values that have not yet been
+        turned into a mask (0s and 1s)
+    masked_values: list
+        A list of cloud mask values
```

Comment:
I think an explanation on where this list should come from maybe helpful.

Reply:
Good idea, thank you!